host/windows/syscall.c

```diff
@@ -366,34 +837,213 @@ oe_host_fd_t oe_syscall_open_ocall(
+        }
+
+        switch (flags & (OE_O_CREAT | OE_O_EXCL | OE_O_TRUNC))
+        {
+            case OE_O_CREAT:
+            {
+                // Create a new file or open an existing file.
+                create_dispos = OPEN_ALWAYS;
+                break;
+            }
+            case OE_O_CREAT | OE_O_EXCL:
+            case OE_O_CREAT | OE_O_EXCL | OE_O_TRUNC:
+            {
+                // Create a new file, but fail if it already exists.
+                // Ignore `O_TRUNC` with `O_CREAT | O_EXCL`
+                create_dispos = CREATE_NEW;
+                break;
+            }
+            case OE_O_CREAT | OE_O_TRUNC:
+            {
+                // Truncate file if it already exists.
+                create_dispos = CREATE_ALWAYS;
+                break;
+            }
+            case OE_O_TRUNC:
+            case OE_O_TRUNC | OE_O_EXCL:
+            {
+                // Truncate file if it exists, otherwise fail. Ignore O_EXCL
+                // flag.
+                create_dispos = TRUNCATE_EXISTING;
+                break;
+            }
+            case OE_O_EXCL:
+            default:
+            {
+                // Open file if it exists, otherwise fail. Ignore O_EXCL flag.
+                create_dispos = OPEN_EXISTING;
+                break;
+            }
+        }
+
+        // in linux land, we can always share files for read and write unless
+        // they have been opened exclusive
+        share_mode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE;
+        const int ACCESS_FLAGS = 0x3; // Covers rdonly, wronly rdwr
+        switch (flags & ACCESS_FLAGS)
+        {
+            case OE_O_RDONLY:
+            {
+                desired_access = GENERIC_READ;
+                break;
+            }
+            case OE_O_WRONLY:
+            {
+                desired_access =
+                    (flags & OE_O_APPEND) ? FILE_APPEND_DATA : GENERIC_WRITE;
+                break;
+            }
+            case OE_O_RDWR:
+            {
+                desired_access =
+                    GENERIC_READ |
+                    ((flags & OE_O_APPEND) ? FILE_APPEND_DATA : GENERIC_WRITE);
+                break;
+            }
+            default:
+                ret = -1;
+                _set_errno(OE_EINVAL);
+                goto done;
+                break;
+        }
+
+        if (mode & OE_S_IRUSR)
+            desired_access |= GENERIC_READ;
+        if (mode & OE_S_IWUSR)
+            desired_access |= GENERIC_WRITE;
+
+        HANDLE h = CreateFileW(
+            wpathname,
+            desired_access,
+            share_mode,
+            NULL,
+            create_dispos,
+            file_flags,
+            NULL);
+        if (h == INVALID_HANDLE_VALUE)
+        {
+            _set_errno(_winerr_to_errno(GetLastError()));
+            goto done;
+        }
+
+        ret = (oe_host_fd_t)h;
+
+        // Windows doesn't do mode in the same way as linux. We can set user
```

Comment:
I don't think this is the right way to do this in Windows; notably `CreateFileW()` supports specifying `SECURITY_ATTRIBUTES` that can specify DACLs for file objects in Windows, and that's where you set the RWX as ACL associal with the file, not as file attributes.

This is non-trivial though, an a mapping of mode to Windows permissions should probably be abstracted out into its own helper functions and source files. For reference, see [Creating a security descriptor for a new object in C](https://docs.microsoft.com/en-us/windows/win32/secauthz/creating-a-security-descriptor-for-a-new-object-in-c--?redirectedfrom=MSDN)

Reply:
@jhand2 @anakrish @nonpolarity 
Design question: Do we actually want to implement correct RWX semantics for syscalls on Windows? It looks like based on the old Windows POSIX wrapper functions that are being called in this implementation that they just did a crappy mapping to the Windows read-only file attribute. If not, this is another area that definitely needs public documentation to call out the unsupported behaviors.

Reply:
I think that supporting RWX is actually not too challenging if we just pass `SECURITY_ATTRIBUTES` to `CreateFileW` as you say. So unless there's something more complicated that I'm missing (which is very possible), I think we should try to implement file permissions properly.

Reply:
Directory is set through SetNamedSecurityInfo(), and file is set with SECURITY_ATTRIBUTES, thanks.